[KOGITO-4889] - Proposal: spliting integration and unit tests

ricardozanini opened the pull request:
Many thanks for submitting your Pull Request :heart:! 

See: https://issues.redhat.com/browse/KOGITO-4889

In this PR we introduce a new profile `quick-build`:

```
mvn clean install -Dquickly
```

It will run **only** unit tests and formatting validation. It takes **7 minutes** to run from the parent `kogito-runtimes` project (was 30 min before).

Integration tests are segregated and now only run on `integration-test` or `verify` phases. Ideally, CI could run first with `-Dquickly`, and then the full cycle with integration tests:

```
# phase 1
mvn clean install -Dquickly

# phase 2 (install has already run)
mvn verify
```

Next steps:

- [x] Peer review consolidation
- [ ] Jenkins pipeline review (ideally IT would run separately, so we can replay in case of infra failure)

In the future we may consider [classifying ITs](https://junit.org/junit5/docs/current/user-guide/#running-tests-tag-expressions), so it would make it easier for us to replay specific modules/scenarios.

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.md)
- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/master/kogito-ide-config)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>Pull Request</b>  
  Please add comment: <b>Jenkins retest this</b>
 
* <b>Quarkus LTS checks</b>  
  Please add comment: <b>Jenkins run LTS</b>

* <b>Native checks</b>  
  Please add comment: <b>Jenkins run native</b>

* <b>Full Kogito testing</b> (with cloud images and operator BDD testing)  
  Please add comment: <b>Jenkins run BDD</b>  
  <b>This check should be used only if a big change is done as it takes time to run, need resources and one full BDD tests check can be done at a time ...</b>
</details>

radtriste reviewed (commented):
I like that PR :) Please consider my suggestion in comment.

Gonna PR on this to update the Jenkinsfile(s)

ricardozanini commented:
Many thanks for your suggestion, @radtriste.

Now, by running `mvn clean install -Dquickly`, the entire build takes **~1min**. With tests enabled (`-DquickTests`), everything is finished in **~7min**. 

Some tests in the `jbpm-flow` module are taking too long because of the test case's sleeps. I tried to fork and run and parallel, but no luck. Forking broke the test, and honestly, I don't have too much time to take a look.

We can try to run some modules in parallel, though. But that will be an improvement for another JIRA. :)

radtriste commented on CONTRIBUTING.md:
```diff
@@ -147,21 +147,23 @@ On Linux, check [the post-installation guide](https://docs.docker.com/install/li
 
-By removing the flags, you will run the corresponding tests. 
-It will take much longer to build but will give you more guarantees on your code. 
+By removing the flags, you will run the unit and integration tests. 
```
```suggestion
By removing the `quickly` property, you will run the whole unit and integration tests. 
```

radtriste commented:
Jenkins retest this

radtriste commented:
Jenkins retest this

ricardozanini commented:
Jenkins retest this

cristianonicolai commented:
jenkins please retest this

radtriste commented:
jenkins please retest this

mbiarnes commented:
Jenkins retest this

evacchi reviewed (approved):
love it ! I wish we could add some sort of Maven enforcer rule to avoid mistakingly adding ITs to Units... I think we should be able to do something if we are creative enough

evacchi commented:
@elguardian you may want to keep an eye on this, since you are refactoring test cases -- not sure if it may have an impact

ricardozanini commented:
Can we ignore Sonar? I haven't change a single line of code. :thinking:

elguardian commented:
@evacchi for now that is frozen... till finishing marshalling

cristianonicolai commented:
jenkins please retest this

ricardozanini commented:
Rebased, IT passed locally.

@radtriste, wanna take a final look?
@jstastny-cz, I've incorporated your changes. Can you please take a look? How do we validate?

radtriste reviewed (approved):
lgtm

cristianonicolai commented:
jenkins please retest this

ricardozanini commented:
The former run failed because of docker. The current CI run is stuck in the containers test again for the past 5 hours.

jstastny-cz commented on addons/persistence/infinispan-persistence-addon/pom.xml:
```diff
@@ -113,19 +113,11 @@
-          </execution>
-        </executions>
+        <configuration>
+          <systemPropertyVariables>
```
```suggestion
          <systemPropertyVariables combine.children="append">
```
This makes sure `maven.repo.local` and settings.xml configuration is passed through to quarkus-based tests.

jstastny-cz commented on kogito-build-parent/pom.xml:
```diff
@@ -1081,18 +1081,29 @@
+                <goal>integration-test</goal>
+                <goal>verify</goal>
+              </goals>
+              <configuration>
```
Move the configuration back into plugin level, that way it will be inherited by all executions, this way, you're reducing it just to the matching id execution only IMO.

jstastny-cz commented on kogito-build-parent/pom.xml:
```diff
@@ -1519,6 +1530,80 @@ limitations under the License.
+            <plugin>
+              <artifactId>maven-invoker-plugin</artifactId>
+              <configuration>
+                <skipInstallation>true</skipInstallation>
```
wouldn't be possible to expose all these skips as properties, set in default pluginManagement `<skipInstallation>${skip.invoker.installation}</skipInstallation>` and then here just set the property value as for formatter.skip, etc ?

ricardozanini replied:
I don't think so. This profile is meant to skip everything. There's no point in exposing these properties. If one need to run with a particular set of properties to skip, they can do it without this profile.

jstastny-cz replied:
What I meant is in general pluginManagement section for each plugin to define these configuration properties (instead of redefining the plugin in here) and this kind of profile then would just define properties ( as already done for all those properties above - formatter.skip, skipITs, invoker.skip, etc. 
We can see that we're actually re-defining skipInvocation (line below) even though we're setting invoker.skip above in properties to true. So there's already a mismatch in the approach in this profile (redefined vs rely on a property).

ricardozanini replied:
We need that otherwise won't work. I couldn't debug the actual plugin. If you figured, we can change it later.

jstastny-cz replied:
I've created ricardozanini/kogito-runtimes#3 to illustrate what I meant.

ricardozanini replied:
I applied your suggestions, but without redeclaring the `<skip>` block in the `quick-test` profile the invoker plugin stills run:

```
[INFO] --- maven-invoker-plugin:3.2.1:run (integration-test) @ integration-tests-springboot ---
[INFO] Building: integration-tests-springboot-norest-it/pom.xml
[INFO] [INFO] Scanning for projects...
```

jstastny-cz reviewed (changes requested):
Please see a few comments.

Regarding to Sonar, please don't ignore that, I would encourage for closer assessment of what this refactoring means towards reported metrics. By moving test into IT phase it might be that it's not included in the report. In addition to traditional problems, quarkus brings some limitations as documented in https://quarkus.io/guides/tests-with-coverage.
Having said that, we have coverage broken anyway, fixing in #1209 .

ricardozanini commented:
> Regarding to Sonar, please don't ignore that, I would encourage for closer assessment of what this refactoring means towards reported metrics. 

That's funny. I just moved the code from one package to another, and Sonar complained because of the new code? That's it? But it was an ok change in a code with more than 11 years. But you're right, we need to take the opportunity to raise our quality level.

I think for this kind of scenario, we should fix the first time we forked or at least opened JIRAs to tackle the technical debt. Fixing this kind of bug in PRs like this is out of scope, IMO.

ricardozanini commented:
> Regarding to Sonar, please don't ignore that, I would encourage for closer assessment of what this refactoring means towards reported metrics. 

That's funny. I just moved the code from one package to another, and Sonar complained because of the new code? That's it? But it was an ok change in a code with more than 11 years. But you're right, we need to take the opportunity to raise our quality level.

I think for this kind of scenario, we should fix the first time we forked or at least opened JIRAs to tackle the technical debt. Fixing this kind of bug in PRs like this is out of scope, IMO.

jstastny-cz commented:
@ricardozanini yeah, I was nitpicking on the word "ignore" :-) if we review and add possible future task in Jira it's I think just about what we can do as part of this effort.

ricardozanini commented:
Seems like we had some kind of connectivity problems?

```
17:17:55,512 [INFO] -------------------------------------------------------
17:17:55,512 [INFO]  T E S T S
17:17:55,512 [INFO] -------------------------------------------------------
17:17:56,522 [INFO] Running org.kie.kogito.integrationtests.quarkus.TopicInformationResourceIT
Cannot contact kie-psi-rhel7-xlarge-2606: java.lang.InterruptedException
```

@mareknovotny @mbiarnes FYI ^

Jenkins retest this please

ricardozanini commented:
> @ricardozanini yeah, I was nitpicking on the word "ignore" :-) if we review and add possible future task in Jira it's I think just about what we can do as part of this effort.

No problem! Can you please review it then? :)

ricardozanini commented:
Errors:

```
10:50:05.898 [main] ERROR ?.k.k.0.0-SNAPSHOT] - Failed to pull image: org.kie.kogito/trusty-service-infinispan:2.0.0-SNAPSHOT. Please check output of `docker pull org.kie.kogito/trusty-service-infinispan:2.0.0-SNAPSHOT`
com.github.dockerjava.api.exception.InternalServerErrorException: Status 500: {"message":"Get https://org.kie.kogito/v2/: dial tcp: lookup org.kie.kogito on 10.11.5.19:53: no such host"}

	at org.testcontainers.shaded.com.github.dockerjava.core.DefaultInvocationBuilder.execute(DefaultInvocationBuilder.java:247) ~[testcontainers-1.15.1.jar:na]
	at org.testcontainers.shaded.com.github.dockerjava.core.DefaultInvocat
```

Not sure if it's related to this PR. @radtriste @r00ta can you guys help?

ricardozanini commented:
Jenkins retest this please

cristianonicolai commented:
jenkins please retest this

radtriste commented:
jenkins please retest this

ricardozanini commented:
Timeout on SB Integration tests :face_with_head_bandage: 

jenkins please retest this

ricardozanini commented:
Stuck again:

```
0:24:37,765 [INFO] Running org.kie.kogito.integrationtests.quarkus.OASIT
```

ricardozanini commented:
jenkins please retest this

radtriste commented:
jenkins please retest this

radtriste commented:
jenkins retest this

ricardozanini commented:
> Caused by: org.testcontainers.containers.ContainerFetchException: Can't get Docker image: RemoteDockerImage(imageName=org.kie.kogito/trusty-service-infinispan:2.0.0-SNAPSHOT, imagePullPolicy=DefaultPullPolicy())

jenkins retest this

cristianonicolai commented:
jenkins please retest this

cristianonicolai commented:
jenkins please retest this

cristianonicolai commented:
jenkins please retest this

cristianonicolai commented:
jenkins please retest this

cristianonicolai commented:
jenkins please retest this

danielezonca commented:
jenkins retest this please

cristianonicolai commented:
jenkins please retest this

cristianonicolai commented:
jenkins please retest this

danielezonca commented:
Nice
`Status 500: {"message":"toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit"}`

jenkins retest this please

danielezonca commented:
Network error
`08:23:02,526 [INFO] [ERROR]     Non-resolvable import POM: Could not transfer artifact io.quarkus:quarkus-bom:pom:1.13.3.Final from/to mirror-central (https://bxms-qe.rhev-ci-vms.eng.rdu2.redhat.com:8443/nexus/content/groups/kogito-runtimes-pr-artifacts-group/): Connect to bxms-qe.rhev-ci-vms.eng.rdu2.redhat.com:8443 [bxms-qe.rhev-ci-vms.eng.rdu2.redhat.com/10.0.13.239] failed: Connection refused (Connection refused) @ org.kie.kogito:kogito-quarkus-bom:2.0.0-SNAPSHOT, /home/jenkins/workspace/KIE/kogito/pullrequest/kogito-runtimes/integration-tests/integration-tests/integration-tests-kogito-plugin/target/local-repo/org/kie/kogito/kogito-quarkus-bom/2.0.0-SNAPSHOT/kogito-quarkus-bom-2.0.0-SNAPSHOT.pom, line 20, column 19 -> [Help 2]`

jenkins retest this please

cristianonicolai commented:
jenkins please retest this

cristianonicolai commented:
jenkins please retest this

cristianonicolai commented:
jenkins please retest this

cristianonicolai commented:
jenkins please retest this

ricardozanini commented:
@danielezonca we are green on GHA side. :)
